recipe-server/client/control/components/recipes/AbstractFormPage.js

```diff
@@ -0,0 +1,77 @@
+import autobind from 'autobind-decorator';
+import PropTypes from 'prop-types';
+import React from 'react';
+
+@autobind
+export default class AbstractFormPage extends React.PureComponent {
+  static propTypes = {
+    recipeAction: PropTypes.func.isRequired,
+  };
+
+  state = {
+    formErrors: undefined,
+  };
+
+  /* hook */ onBeforeSubmit() {
+    this.setState({
+      formErrors: undefined,
+    });
+  }
+
+  /* abstract */ onFailure() {
+    throw new Error('AbstractFormPage#onFailure should be overridden.');
+  }
+
+  /* abstract */ onSuccess() {
+    throw new Error('AbstractFormPage#onSuccess should be overridden.');
+  }
+
+  /* hook */ getFormProps() {
+    return {};
+  }
+
+  /* abstract */ getFormComponent() {
+    throw new Error('AbstractFormPage#getFormComponent should be overridden.');
+  }
+
+  /* abstract */ getTitle() {
+    throw new Error('AbstractFormPage#getTitle should be overridden.');
+  }
+
+  /* abstract */ async performAction() {
+    throw new Error('AbstractFormPage#onSuccess should be overridden.');
+  }
+
+  /**
+   * Update the existing recipe and display a message.
+   */
+  async handleSubmit(formValues) {
+    try {
+      const actionValues = await this.performAction(formValues);
+
+      this.onSuccess(...actionValues);
```

Comment:
I'd like to explore this a bit so I haven't touched it yet. The idea behind the spread operator was to keep functions written as `doThing(someVal, otherVal)` versus `doThing(vals){ someVal = vals[0]; otherVal = vals[1]; }`. I will test out some ideas and go from there.

Reply:
Without spread, the second example could be written as `doThings([someVal, otherVal])`. It would also let users do things like `doThings({ someVal, otherVal })` (ie, passing an options object instead of ordered arguments), or a simple single value `doThings(someVal)`, neither of which would be allowed with the spread operator.